### DISCUSSION TITLE
Unicode support in `Paragraph` and elsewhere

### DISCUSSION THREAD

**Discussion by lthoerner** [ratatui/ratatui#192] [Discussion]
It seems to me that implementing proper Unicode support and representing every character as monospace are two fundamentally-incompatible aims. Where are we drawing the line as to where predictability of char display takes precedent over locale support?

I bring this up because I am trying to implement hard-wrapping, but because Unicode isn't exactly based on chars, it seems to be extremely difficult, if not impossible, to allow users to predict where text will be displayed (which in my own experience is crucial in some cases).

Note: it may be a good idea to check whether there are any crates currently using `ratatui` that need international locale support in the first place.

**Comment by joshka** [comment]
You might have seen this already, but I'm calling it out as it shows two things - people expect unicode to work, and unicode can be difficult even when you do solve the issue at the tui layer. Your terminal can be a problem too.
- https://github.com/tui-rs-revival/ratatui/issues/75

> "Unicode ... Where are we drawing the line"

I see what you did there... :D

> Note: it may be a good idea to check whether there are any crates currently using ratatui that need international locale support in the first place.

Unicode is more than just locale. It's emoji and other symbols.
E.g.: https://crates.io/crates/xplr uses symbols when drawing the tree
See image at https://xplr.stck.me/post/25252/What-is-a-TUI-file-explorer-why-would-you-need-one

I'd imagine people using twitch-tui would probably not be enthused if we dropped emoji support too.

Perhaps a reasonable approach to solving this would be to make it easy to pass in an implementation that handles the wrapping rather than baking in approaches. So if you want an unicode aware wrapping then you pass in the unicode wrapper that doesn't worry so much about single character cells.
